c7n/resources/s3.py

```diff
@@ -1191,10 +1191,15 @@ def process_key(self, s3, key, bucket_name, info=None):
         if info is None:
             info = s3.head_object(Bucket=bucket_name, Key=k)
 
-        if 'ServerSideEncryption' in info:
-            if self.kms_id and info.get('SSEKMSKeyId', '') == self.kms_id:
-                return False
-            else:
+        # If the data is already encrypted with AES256 and this request is also
+        # for AES256 then we don't need to do anything
+        if info.get('ServerSideEncryption') == 'AES256' and not self.kms_id:
+            return False
+
+        # If the data is already encrypted with KMS and the same key is provided
+        # then we don't need to do anything
+        if info.get('ServerSideEncryption') == 'aws:kms' and self.kms_id:
+            if self.kms_id in info.get('SSEKMSKeyId', ''):
```

Comment:
`in` is intentional.  The original code was buggy because it was using equality, although it returned False no matter what so the conditional didn't matter anyways.

self.kms_id looks like this: `dacca8ea-f166-43af-aaaa-b35f8781035d`

SSEKMSKeyId will look like this: `arn:aws:kms:us-east-1:xxxxxxxxxxx:key/dacca8ea-f166-43af-aaaa-b35f8781035d`

I could split on the forward slash if you think that is a better choice, but chose `in` because the code read easier.